From: Gennaro Oliva <oliva.g@na.icar.cnr.it>
Date: Fri, 5 Sep 2025 17:04:25 -0600
Subject: Retry getaddrinfo five times before giving up

Forwarded: https://bugs.schedmd.com/show_bug.cgi?id=13280
Last-Update: 2024-12-01

This patch retry getaddrinfo five times before giving up and exiting SLURM
deamons. This is needed on Debian systems when using ifupdown for network
configuration with allow-hotplug interfaces (which the installer uses by
default). This settings do not guarantee that getaddrinfo succeed after
network-online.target is complete.
---
 src/common/util-net.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/common/util-net.c b/src/common/util-net.c
index 549d609..3858b8e 100644
--- a/src/common/util-net.c
+++ b/src/common/util-net.c
@@ -39,6 +39,7 @@
 
 #define _GNU_SOURCE
 
+#include <unistd.h>
 #include <arpa/inet.h>
 #include <errno.h>
 #include <limits.h>	/* for PATH_MAX */
@@ -258,8 +259,15 @@ static struct addrinfo *_xgetaddrinfo(const char *hostname, const char *serv,
 {
 	struct addrinfo *result = NULL;
 	int err;
+	int retry = 10;
+
+	while ( ( (err = getaddrinfo(hostname, serv, hints, &result)) != 0 ) && retry ) {
+		error("%s: getaddrinfo() failed: %s: %m, attempt number %d", __func__,
+			gai_strerror(err), 6 - retry);
+		sleep(1);
+		retry -= 1;
+	}
 
-	err = getaddrinfo(hostname, serv, hints, &result);
 	if (err == EAI_SYSTEM) {
 		error_in_daemon("%s: getaddrinfo(%s:%s) failed: %s: %m",
 				__func__, hostname, serv, gai_strerror(err));
